src/apps/replication/lib/replication_failure_detector.cpp

```diff
@@ -62,71 +62,93 @@ replication_failure_detector::~replication_failure_detector(void)
+    }
+    if ( cur_record.last_send_time_for_beacon_with_ack > ack.time) {
+        dinfo("out dated beacon, ignore");
+        return;
+    }
+    if (err==ERR_OK && !ack.allowed) {
+        switch_master_accordingly(err, ack);
         return;
     }
 
+    if ( !cur_record.is_alive )
     {
-        zauto_lock l(_meta_lock);
-
-        if (dsn_group_is_leader(_meta_servers.group_handle(), ack.this_node.c_addr()))
+        if (err==ERR_OK && ack.is_master)
+            failure_detector::end_ping_internal(err, ack);
+        else
+            switch_master_accordingly(err, ack);
+    }
+    else
+    {
+        if (err!=ERR_OK)
         {
-            if (err != ERR_OK)
-            {
-                // ping failed, switch master
-                rpc_address node = dsn_group_next(_meta_servers.group_handle(), ack.this_node.c_addr());
-                if (ack.this_node != node)
-                {
-                    dsn_group_set_leader(_meta_servers.group_handle(), node.c_addr());
-                    switch_master(ack.this_node, node);
-                }
-            }
-            else if (ack.is_master == false)
-            {
-                // switch master
-                if (!ack.primary_node.is_invalid()
-                        && !dsn_group_is_leader(_meta_servers.group_handle(), ack.primary_node.c_addr()))
-                {
-                    dsn_group_set_leader(_meta_servers.group_handle(), ack.primary_node.c_addr());
-                    switch_master(ack.this_node, ack.primary_node);
-                }
-            }
-            else
-            {
-            }
+            /*
+             * we ignore the timeout messages.
+             * Basicly, we rely on the "on_master_disconnected",
+             * then we will switch to another master
+             */
+            return;
         }
-        else // ack.this_node is not leader
+        else
         {
-            if (err != ERR_OK)
-            {
-                // do nothing
-            }
-            else if (ack.is_master == false)
-            {
-                // switch master
-                if (!ack.primary_node.is_invalid()
-                        && !dsn_group_is_leader(_meta_servers.group_handle(), ack.primary_node.c_addr()))
-                {
-                    dsn_group_set_leader(_meta_servers.group_handle(), ack.primary_node.c_addr());
-                    switch_master(ack.this_node, ack.primary_node);
+            if (ack.is_master)
+                failure_detector::end_ping_internal(err, ack);
+            else {
+                /*
+                 * Well, we are now connected with a master happily. But he suddenly says
+                 * he is no leader any more.
+                 * in this case, we need to connect to a new one
+                 */
+                if ( !ack.primary_node.is_invalid() )
+                    switch_master_accordingly(err, ack);
+                else {
+                    cur_record.is_alive = false;
```

Comment:
unnecessary disconnection may be caused even the switch happens within the lease period